### PR TITLE
Resync trie when get node from db fails

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

File: data/trie/node.go [hunk 1]
```diff
@@ -97,7 +98,7 @@ func encodeNodeAndCommitToDB(n node, db data.DBWriteCacher) error {
 func getNodeFromDBAndDecode(n []byte, db data.DBWriteCacher, marshalizer marshal.Marshalizer, hasher hashing.Hasher) (node, error) {
 	encChild, err := db.Get(n)
 	if err != nil {
-		return nil, fmt.Errorf("getNodeFromDB error %w for key %v", err, hex.EncodeToString(n))
+		return nil, fmt.Errorf(core.GetNodeFromDBErrorString+"%w for key %v", err, hex.EncodeToString(n))
```
Comment (iulianpascalau) [inline data/trie/node.go:101]: add an empty space before the "%"

File: process/sync/metablock.go [hunk 2]
```diff
@@ -152,7 +165,44 @@ func (boot *MetaBootstrap) setLastEpochStartRound() {
 // These methods will execute the block and its transactions. Finally if everything works, the block will be committed
 // in the blockchain, and all this mechanism will be reiterated for the next block.
 func (boot *MetaBootstrap) SyncBlock() error {
-	return boot.syncBlock()
+	err := boot.syncBlock()
+
```
Comment (iulianpascalau) [inline process/sync/metablock.go:169]: remove empty line

File: process/sync/metablock.go [hunk 3]
```diff
@@ -152,7 +165,44 @@ func (boot *MetaBootstrap) setLastEpochStartRound() {
...
+	if err != nil && strings.Contains(err.Error(), core.GetNodeFromDBErrorString) {
```
Comment (iulianpascalau) [inline process/sync/metablock.go:170]: this could have been a bool var

File: process/sync/metablock.go [hunk 4]
```diff
@@ -152,7 +165,44 @@ func (boot *MetaBootstrap) setLastEpochStartRound() {
...
+		}
+	}
+
+	return err
+}
+
+func (boot *MetaBootstrap) syncAccountsDBs() error {
+	var err error
+
+	log.Debug("base sync: started syncPeerAccountsState")
```
Comment (iulianpascalau) [inline process/sync/metablock.go:183]: maybe log.Warn? as these should not normally happen.

File: process/sync/shardblock.go [hunk 5]
```diff
@@ -127,7 +129,16 @@ func (boot *ShardBootstrap) StartSyncingBlocks() {
 // These methods will execute the block and its transactions. Finally if everything works, the block will be committed
 // in the blockchain, and all this mechanism will be reiterated for the next block.
 func (boot *ShardBootstrap) SyncBlock() error {
-	return boot.syncBlock()
+	err := boot.syncBlock()
+
```
Comment (iulianpascalau) [inline process/sync/shardblock.go:133]: remove empty line

File: process/sync/shardblock.go [hunk 6]
```diff
@@ -127,7 +129,16 @@ func (boot *ShardBootstrap) StartSyncingBlocks() {
...
+	if err != nil && strings.Contains(err.Error(), core.GetNodeFromDBErrorString) {
```
Comment (iulianpascalau) [inline process/sync/shardblock.go:134]: this could have been a bool var

**Review by iulianpascalau — COMMENTED** [review]
minor stuff

File: data/trie/node.go (hunk 1 above)
Comment (sasurobert) [reply data/trie/node.go:101]: done

File: process/sync/metablock.go (hunk 2 above)
Comment (sasurobert) [reply process/sync/metablock.go:169]: done

File: process/sync/metablock.go (hunk 3 above)
Comment (sasurobert) [reply process/sync/metablock.go:170]: done

File: process/sync/shardblock.go (hunk 5 above)
Comment (sasurobert) [reply process/sync/shardblock.go:133]: done

File: process/sync/shardblock.go (hunk 6 above)
Comment (sasurobert) [reply process/sync/shardblock.go:134]: done

**Review by iulianpascalau — DISMISSED** [review]

**Review by iulianpascalau — DISMISSED** [review]

File: process/sync/baseSync.go [hunk 7]
```diff
@@ -624,6 +628,17 @@ func (boot *baseBootstrap) syncBlock() error {
 	return nil
 }
 
+func (boot *baseBootstrap) syncUserAccountsState() error {
+	rootHash, err := boot.accounts.RootHash()
+	if err != nil {
+		log.Error("SyncBlock syncUserAccountsState", "error", err)
```
Comment (bogdan-rosianu) [inline process/sync/baseSync.go:634]: this error is logged also on the caller

File: process/sync/metablock.go [hunk 8]
```diff
@@ -157,7 +170,43 @@ func (boot *MetaBootstrap) setLastEpochStartRound() {
...
+		return err
+	}
+
+	return nil
+}
+
+func (boot *MetaBootstrap) syncValidatorAccountsState() error {
+	rootHash, err := boot.validatorAccountsDB.RootHash()
+	if err != nil {
+		log.Error("SyncBlock syncUserAccountsState", "error", err)
```
Comment (bogdan-rosianu) [inline process/sync/metablock.go:204]: this error is logged also on the caller

File: process/sync/metablock.go [hunk 9]
```diff
@@ -157,7 +170,43 @@ func (boot *MetaBootstrap) setLastEpochStartRound() {
 // These methods will execute the block and its transactions. Finally if everything works, the block will be committed
 // in the blockchain, and all this mechanism will be reiterated for the next block.
 func (boot *MetaBootstrap) SyncBlock() error {
-	return boot.syncBlock()
+	err := boot.syncBlock()
+	isErrGetNodeFromDB := err != nil && strings.Contains(err.Error(), core.GetNodeFromDBErrorString)
```
Comment (bogdan-rosianu) [inline process/sync/metablock.go:174]: extract L174-L180 to a function `syncAccountsIfNeeded(err error)`

File: process/sync/shardblock.go [hunk 10]
```diff
@@ -132,7 +134,16 @@ func (boot *ShardBootstrap) StartSyncingBlocks() {
 // These methods will execute the block and its transactions. Finally if everything works, the block will be committed
 // in the blockchain, and all this mechanism will be reiterated for the next block.
 func (boot *ShardBootstrap) SyncBlock() error {
-	return boot.syncBlock()
+	err := boot.syncBlock()
+	isErrGetNodeFromDB := err != nil && strings.Contains(err.Error(), core.GetNodeFromDBErrorString)
```
Comment (bogdan-rosianu) [inline process/sync/shardblock.go:138]: extract L138-L144 to a function `syncAccountsIfNeeded(err error)`

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: process/sync/metablock.go (hunk 8 above)
Comment (sasurobert) [reply process/sync/metablock.go:204]: deleted

File: process/sync/baseSync.go (hunk 7 above)
Comment (sasurobert) [reply process/sync/baseSync.go:634]: deleted

File: factory/consensusComponents.go [hunk 11]
```diff
@@ -449,6 +463,51 @@ func (ccf *consensusComponentsFactory) createShardBootstrapper() (process.Bootst
...
+		Timeout:                   timeoutGettingTrieNode,
+		Cacher:                    ccf.dataComponents.Datapool().TrieNodes(),
+		MaxTrieLevelInMemory:      ccf.config.StateTriesConfig.MaxStateTrieLevelInMemory,
+		MaxHardCapForMissingNodes: ccf.config.TrieSync.MaxHardCapForMissingNodes,
+		TrieSyncerVersion:         ccf.config.TrieSync.TrieSyncerVersion,
+	}
+}
+
+func (ccf *consensusComponentsFactory) createValidatorAccountsSyncer() (process.AccountsDBSyncer, error) {
+	trieStorageManager, ok := ccf.stateComponents.TrieStorageManagers()[factory.PeerAccountTrie]
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply factory/consensusComponents.go:481]: it should never return nil.

File: process/sync/metablock.go (hunk 9 above)
Comment (sasurobert) [reply process/sync/metablock.go:174]: extraction does not work here. it complicates things. needs internal interface and all.

File: process/sync/shardblock.go (hunk 10 above)
Comment (sasurobert) [reply process/sync/shardblock.go:138]: extraction does not work here. it complicates things. needs internal interface and all.

**Review by bogdan-rosianu — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Review by iulianpascalau — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.
image

**Pull request merged**